### PR TITLE
Court service and facade basics

### PR DESCRIPTION
**Pull request by paveltobias**

### CODE REVIEW

File: api/src/main/java/tennisclub/facade/CourtFacade.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+import tennisclub.dto.CourtDto;
+import tennisclub.enums.CourtType;
+import java.util.List;
+
+/**
+ * @author Pavel Tobiáš
+ */
+public interface CourtFacade {
+    void create(CourtDto court);
```
Comment (Fuyunori) [inline api/src/main/java/tennisclub/facade/CourtFacade.java:11]: I am not sure whether the argument should be `CourtDto`. When I looked at the reference implementation, they had another DTO _(in your case it would be a `CourtCreateDTO` - it is almost identical to the CourtDTO, but it lacks the `id` attribute because when you create a new instance, you create it without its `id`)_. Personally, I am not very fond of the DTO pattern so personally I would approve but I am not sure whether the seminar tutor would like it 😁.

**Review by Fuyunori — COMMENTED** [review]
Thanks for moving the enums into the common module! 😊

File: api/src/main/java/tennisclub/facade/CourtFacade.java (hunk 1 above)
Comment (paveltobias) [reply api/src/main/java/tennisclub/facade/CourtFacade.java:11]: @Fuyunori , let me add it then. Do you think would be a problem if `CourtDto extends CourtCreateDto` to reduce the amount of boilerplate code?

Comment (Fuyunori) [reply api/src/main/java/tennisclub/facade/CourtFacade.java:11]: @paveltobias I was actually researching today whether it is good to have a DTO object inherit from another. I haven't seen any convincing answers, however based on what I have seen, most of them discourage from doing so _(but again, personally I would also prefer inheritance in this case)_. So I myself don't really know answer to this question - we can ask about this on the seminar 😁

Comment (paveltobias) [reply api/src/main/java/tennisclub/facade/CourtFacade.java]: @Fuyunori, oops, already made the change using inheritance. :laughing: Never mind, let me correct it. :smile:

Comment (Fuyunori) [reply api/src/main/java/tennisclub/facade/CourtFacade.java]: @paveltobias you can leave it as it is 😁😀

Comment (paveltobias) [reply api/src/main/java/tennisclub/facade/CourtFacade.java]: @Fuyunori, lol, we are somewhat out of sync; already did the fix. :laughing: Anyhow, the PR should be mergeable now.

Comment (paveltobias) [reply api/src/main/java/tennisclub/facade/CourtFacade.java]: @Fuyunori, ...but let us ask the tutor anyway. Do you still have links to the sources discouraging from using inheritance between DTOs?

Comment (Fuyunori) [reply api/src/main/java/tennisclub/facade/CourtFacade.java]: @paveltobias Thanks and good job 😊👏 Also I think the DTO might require having `equals` and `hashCode` _(or at least the reference implementation has it, but I'm not sure what is the correct approach there)_. 😁

Comment (Fuyunori) [reply api/src/main/java/tennisclub/facade/CourtFacade.java]: @paveltobias this is the probably the most convincing one: https://stackoverflow.com/questions/37870918/should-dtos-use-inheritance-or-composition - especially this sentence:

> The "one DTO class" approach is almost certainly bad. It smells like a God Class. Many pundits excoriate DTOs altogether. You could inherit from some base class, but for value objects it's not really sensible. Same for composition. It makes your code more complicated.

Because it's a value class, it just simply doesn't make much sense to use inheritance 😐

Comment (Fuyunori) [reply api/src/main/java/tennisclub/facade/CourtFacade.java]: @paveltobias there is another link: https://softwareengineering.stackexchange.com/questions/399227/is-it-anti-pattern-to-have-inheritence-in-a-dto

> It sounds kind of funny to me when I read it out loud. Inheritance in DTOs becomes problematic, because you introduce coupling between classes where each class usually represents a concrete idea or use case. DTOs sharing a common parent class implies the class hierarchy will evolve at the same time, and for the same reasons. This makes these classes harder to change without having a ripple effect cascading to classes where the new fields might not be applicable.

Comment (Fuyunori) [reply api/src/main/java/tennisclub/facade/CourtFacade.java]: @paveltobias semantically it doesn't make much sense _(CourtDTO IS NOT a CourtCreateDTO)_ - doesn't comply with the is-A relationship _(inheritance)_.

Comment (paveltobias) [reply api/src/main/java/tennisclub/facade/CourtFacade.java]: @Fuyunori, `equals()` and `hashCode()` generated for both DTOs. :heavy_check_mark:

Comment (paveltobias) [reply api/src/main/java/tennisclub/facade/CourtFacade.java]: @Fuyunori Thank you for the links and arguments! I will read through them and think about it. :smiley:

File: api/src/main/java/tennisclub/dto/CourtDto.java [hunk 2]
```diff
@@ -48,4 +48,27 @@ public String getPreviewImageUrl() {
     public void setPreviewImageUrl(String previewImageUrl) {
         this.previewImageUrl = previewImageUrl;
     }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof CourtDto)) return false;
```
Comment (Fuyunori) [inline api/src/main/java/tennisclub/dto/CourtDto.java:55]: Sorry for being an annoying reviewer 😅in the reference implementation, they stricly don't use `instanceof`, but compare the classes directly like `getClass() != o.getClass()` _(I think it makes sense - either it is the same class or not, you can't get anything else)_ - however, I don't think they ever mentioned this on the lectures or write about this somewhere on the Internet🤔.

Comment (paveltobias) [reply api/src/main/java/tennisclub/dto/CourtDto.java:55]: @Fuyunori, you are **by no means** annoying; thank you for catching my mistakes! Anyway:

- If I remember correctly, the lecture on persistence specifically instructed us to use `instanceof` rather than `getClass()` in our implementations of `equals()`. However, it was mentioned in the context of JPA entities rather than DTOs and might not apply here.
- In any case, if DTOs are never inherited from -- as we discussed earlier -- are `getClass()` and `instanceof` no pretty much equivalent?

Comment (Fuyunori) [reply api/src/main/java/tennisclub/dto/CourtDto.java:55]: @paveltobias Yes, as you said, they just mentioned `equals` and `hashCode` in the context of entity classes 😕As I was not sure in this matter _(again 🤣😆)_, I tried to make a research on `instanceof` and `getClass()` - these links pretty much sums everything 😄 

https://stackoverflow.com/questions/4989818/instanceof-vs-getclass
https://stackoverflow.com/questions/596462/any-reason-to-prefer-getclass-over-instanceof-when-generating-equals

TLDR: you're right, we can actually use `instanceof`, but it is pointless because we don't actually inheritance. I would probably summarize it this way: if you don't use inheritance ➡️  use `getClass` by default, otherwise use `instanceof` 😃

Comment (paveltobias) [reply api/src/main/java/tennisclub/dto/CourtDto.java:55]: @Fuyunori, actually, the sources you have provided seem to suggest that the best solution would be to:
- use `instanceof`,
- mark the `equals()` method `final`.

That way, both:
1. the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle) is maintained (the DTO can be safely replaced by its subclass without altering the method's behavior);
2. the fulfillment of the `equals()` contract is ensured (since `equals()` cannot be overridden, the symmetry of equals cannot be broken by subclasses).

This is, of course, still relevant only if the DTOs will ever be subclassed. By a developer, they should not be (as discussed earlier). But they **could** be _magically_ subclassed under the hood by some of the libraries/frameworks we use. Hibernate does that with the DAOs (as discussed in the lectures); e.g. Mockito, Dozer, etc. **could** do similar _magic_ with the DTOs.

Perhaps, I would stick with the `instanceof` for now (since it cannot break anything) and ask the tutor about his opinion on this [religious debate](https://stackoverflow.com/a/596885). :smile:

Comment (Fuyunori) [reply api/src/main/java/tennisclub/dto/CourtDto.java:55]: @paveltobias that would probably be the best course of action _(to ask the seminar tutor)_ 😄 however I find the `final equals` quite exotic and unusual😆

File: api/src/main/java/tennisclub/dto/CourtCreateDto.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
+package tennisclub.dto;
+
+import tennisclub.enums.CourtType;
+
+public class CourtCreateDto {
```
Comment (Fuyunori) [inline api/src/main/java/tennisclub/dto/CourtCreateDto.java:5]: I've just also noticed one thing from the reference implementation - the XXXCreateDTOs have annotations for validating attributes. So in the case of `Court`, we don't want to have nullable name _(therefore there should be a `@NotNull` annotation; also we don't probably want an empty string as a name, so there could be a `@Size` annotation)_ 😊 I'm sorry I didn't notice it earlier 🙏

File: api/src/main/java/tennisclub/dto/CourtDto.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof CourtDto)) return false;
+
+        CourtDto courtDto = (CourtDto) o;
+
+        if (getName() != null ? !getName().equals(courtDto.getName()) : courtDto.getName() != null) return false;
```
Comment (Fuyunori) [inline api/src/main/java/tennisclub/dto/CourtDto.java:59]: Isn't actually `name` and `address` enough _(as in the entity class)_? 🤔

Comment (paveltobias) [reply api/src/main/java/tennisclub/dto/CourtDto.java:59]: @Fuyunori, `*sigh*` maybe.

- All internet sources suggest that you are right. (I.e. `equals()` should be based on a **small** set of attributes ideally forming a candidate key.) That is what I based my implementation of `CourtDao.equals()` on, as well as my previous comments on (I believe) your past PRs. I also criticized another team in the milestone 1 review for an overblown `equals()`.

-  On the other hand, the seminar's *vox populi* says otherwise. And, ironically, we also got `equals()`-related criticism in the review -- that our `equals()` DAO methods are attribute-wise incomplete. :smile: 

Honestly, I am getting tired of the idiosyncrasies of the Java EE stack and the "religious debates" around it. And I kind of do not care anymore, as long as I get a passing grade, TBH. :frowning_face: 

So... sure, let me *fix* this.

File: api/src/main/java/tennisclub/dto/CourtCreateDto.java (hunk 3 above)
Comment (paveltobias) [reply api/src/main/java/tennisclub/dto/CourtCreateDto.java:5]: Again, thank you for noticing! Will fix.

File: api/src/main/java/tennisclub/dto/CourtDto.java (hunk 4 above)
Comment (paveltobias) [reply api/src/main/java/tennisclub/dto/CourtDto.java:59]: @Fuyunori, done.

File: api/src/main/java/tennisclub/dto/CourtCreateDto.java (hunk 3 above)
Comment (paveltobias) [reply api/src/main/java/tennisclub/dto/CourtCreateDto.java:7]: @Fuyunori, done.

File: api/src/main/java/tennisclub/dto/CourtDto.java (hunk 4 above)
Comment (Fuyunori) [reply api/src/main/java/tennisclub/dto/CourtDto.java:59]: @paveltobias Thank you very much and good job! ❤️

File: api/src/main/java/tennisclub/dto/CourtCreateDto.java (hunk 3 above)
Comment (Fuyunori) [reply api/src/main/java/tennisclub/dto/CourtCreateDto.java:7]: Thank you very much! I think you're just missing also `@NotNull` annotation above the name attribute, and with that it's perfect! 😃 well done and good job :3

**Review by Fuyunori — APPROVED** [review]
I think it's all okay now! ❤️
